src/Shared/diagnostics.ts

```diff
@@ -19,15 +21,37 @@ function issue(id: number) {
 }
 
 let id = 0;
+
 /**
  * Returns a `DiagnosticFactory` that includes a function used to generate a readable message for the diagnostic.
  * @param messages The list of messages to include in the error report.
  */
-function diagnostic(category: ts.DiagnosticCategory, ...messages: Array<string>) {
-	const result = (node: ts.Node) => {
+function diagnostic(category: ts.DiagnosticCategory, ...messages: Array<string>): DiagnosticFactory {
+	return diagnosticWithContext<void>(category, undefined, ...messages);
+}
+
+/**
+ * Returns a `DiagnosticFactory` that includes a function used to generate a readable message for the diagnostic.
+ * The context is additonal data from the location where the diagnostic occurred that is used to generate dynamic
+ * messages.
+ * @param contextFormatter An optional function to format the context parameter for this diagnostic. The returned
+ * formatted messages are displayed last in the diagnostic report.
+ * @param messages The list of messages to include in the diagnostic report.
+ */
+function diagnosticWithContext<T>(
+	category: ts.DiagnosticCategory,
+	contextFormatter?: DiagnosticContextFormatter<T>,
+	...messages: Array<string>
+): DiagnosticFactory<T> {
+	const result = (node: ts.Node, context: T) => {
 		if (category === ts.DiagnosticCategory.Error) {
 			debugger;
 		}
+
+		if (contextFormatter) {
+			messages = messages.concat(contextFormatter(context));
```

Comment:
Any particular reason to use `.concat()` over `messages.push(...contextFormatter(context))`?

Reply:
Nope, no reason other than it slipped my mind. Fixed!